### DISCUSSION TITLE
[Feature Request] Add command to rebuild all installed ports

### DISCUSSION THREAD

**Discussion by sharpe5** [microsoft/vcpkg#17818] [Ideas]
I'd like to rebuild the existing `vcpkg` directory under a newer version of `gcc`.

I'm wondering - is this possible without manually reinstalling everything from scratch?

Not that it matters, but I'm using `WSL v1` with `Ubuntu 18.04`.

**Comment by sharpe5** [comment]
I've searched everywhere in the documentation, I cannot seem to find any mention of this use case.

The reason I am asking is that I suspect that `gcc v9.3.0` has a stability regression with `-O3` compared to `v9.2.1`, and I want to run some tests to verify this.

**Comment by ghost** [comment]
Why not just create a new instance of vcpkg?
- git clone https://github.com/Microsoft/vcpkg.git ./vcpkg-9.3.0
- cd ./vcpkg-9.3.0
- ./bootstrap-vcpkg.sh
- modify the linux triplet to use gcc 9.3.0

I personally have 4 instances, vs2017.prod vs2017.test vs2019.prod and vs2019.test each using the compiler, linker and toolset native to that visual studio version. 

To save some space I did create a common downloads directory so they share the downloaded files, tools, etc.

**Comment by NancyLi1013** [comment]
@sharpe5 
Thanks for posting this issue.

- If you want to rebuild some ports you already installed, you can remove them and then install them again.

- If you want to rebuild vcpkg instance, you can rerun `./bootstrap-vcpkg.sh`.

- Of course, you can clone a new vcpkg as @timautry said above, and then rebuild them.

Please let me know if you still have any concern about this issue.

If there are some problems on` gcc v9.3.0`, please open a new issue to track it.

Thanks.

**Comment by sharpe5** [comment]
@timautry, @NancyLi1013 Thank you for your comments.

It would be fantastic to have a command to rebuild all packages. That way, if the compiler was upgraded (or changed), it would be easy to build everything under the new compiler.

I guess the only solution, for now, is to write some sort of script to remove/install the packages one by one? Not that would be difficult, but it would be a bit hacky.

**Reply by Neumann-A** [reply]
Simply completely remove the `/installed/` dir get a list of installed before doing that by `vcpkg list` use the result of that to reinstall. Manifest mode would automatically rebuild for your project on compiler change.

**Reply by NicolBolas** [reply]
@Neumann-A 

I wouldn't call that a "simply" thing. Some of us have a lot of packages, so we'd have to write a script to parse the result of `vcpkg list` to turn that into a long `vcpkg install` command. Not to mention the added network traffic/time you get by having to re-download what you already had on your drive.

This sounds like the sort of thing the system should be able to handle automatically with a single `vcpkg rebuild` command.

**Reply by Neumann-A** [reply]
>  if the compiler was upgraded

 means you have nothing 

> Not to mention the added network traffic/time you get by having to re-download what you already had on your drive.

I didn't tell you to remove `/downloads/` ....

> vcpkg list to turn that into a long vcpkg install command

use output of `vcpkg list` as response file to `vcpkg install` after removing the description?

**Reply by dg0yt** [reply]
> use output of `vcpkg list` as response file to `vcpkg install` after removing the description?

Maybe it needs a `--brief` option or similar?

**Comment by NancyLi1013** [comment]
@strega-nil 
Could you please help comment about this issue?

**Comment by strega-nil** [comment]
I think this is a pretty good idea; if you wanted to add a command to do it @sharpe5, I could probably offer assistance :)

**Reply by NicolBolas** [reply]
Is there any movement on this issue? The issue was closed recently, but I don't see anything about the feature being added.

Don't people recompile their libraries when they install new versions of the compiler? Wouldn't it be useful to have a single command to do that, rather than having to list all of the installed packages out and reinstall each one manually?

**Reply by JackBoosY** [reply]
No it's not closed, we just convert it to discussion.

**Comment by sharpe5** [comment]
@strega-nil Thank you for the offer, that is very generous. However I don't think I have enough C++ experience yet to add a command like this.

**Comment by strega-nil** [comment]
Fair enough :)

**Comment by sharpe5** [comment]
Please ignore the title change, the change by @strega-nil was correct.

**Comment by gvanem** [comment]
On a similar issue; I now have 7 GByte under my `%VCPKG_ROOT%\downloads\tools` tree. Including 17 Python installations!
Is it safe to just delete everything under it? (and trust it will be redownloaded when needed?).

**Reply by Osyotr** [reply]
Yes.

**Reply by JackBoosY** [reply]
Correct.

**Comment by sketch34** [comment]
Read through all these comments and still don't have a clear picture of how to get vcpkg to recompile everything. Can someone please post manual steps? e.g. Delete folders X, Y, Z...

**Reply by JackBoosY** [reply]
1. you can use command `upgrade` to rebuild all the outdated ports. For the others which not outdated, it should not be rebuilt.
2. Use manifest mode.

**Reply by sketch34** [reply]
I want to rebuild everything not just outdated ports. I'm using manifest mode.

**Reply by JackBoosY** [reply]
Please disable binarycaching then try again.
See https://github.com/microsoft/vcpkg/blob/master/docs/users/binarycaching.md

**Reply by sketch34** [reply]
Thanks! But hmm, running `./vcpkg --binarysource=clear` just shows the command line options to me, as if the command option is not recognised.

I tried renaming my binary cache directory but still the ports did not recompile.

```
[jarrod@xxxx ~]$ cd $HOME/.cache/vcpkg/archives
[jarrod@xxxx archives]$ ls
00  08  0f  16  1c  25  2b  34  3b  46  4f  56  5f  66  6f  7b  81  8b  91  9c  a3  ad  b3  ba  c2  c9  cf  d7  de  e5  ed  f8
01  09  11  17  1f  26  2c  35  3c  48  50  57  60  67  73  7c  82  8c  92  9d  a6  ae  b5  bb  c3  ca  d0  d9  df  e7  ef  f9
03  0a  12  18  20  27  2d  36  3d  49  51  59  61  68  75  7d  83  8d  93  9e  a7  af  b6  bc  c4  cb  d1  da  e0  e9  f0  fb
04  0b  13  19  22  28  2e  38  3e  4a  52  5a  62  6a  77  7e  86  8e  97  9f  a8  b0  b7  bd  c5  cc  d3  db  e1  ea  f2  fe
06  0c  14  1a  23  29  31  39  40  4c  53  5b  63  6b  78  7f  89  8f  98  a1  a9  b1  b8  be  c7  cd  d4  dc  e2  eb  f3  ff
07  0e  15  1b  24  2a  32  3a  42  4d  55  5c  65  6d  7a  80  8a  90  9a  a2  ac  b2  b9  c1  c8  ce  d6  dd  e3  ec  f4
[jarrod@xxxx archives]$ cd ..
[jarrod@xxxx vcpkg]$ mv archives/ archives-bak
```

**Reply by JackBoosY** [reply]
`--binarysource` is a install option, you should append it after `install`.

**Reply by autoantwort** [reply]
What is the reason why you want to rebuild the packages? If your compiler changes or a port is outdated running `vcpkg install` in manifest mode is enough.

**Reply by sketch34** [reply]
@autoantwort I'm getting a linker error for the release build that only happens on my dev machine. The error can be indicative that some dependencies have been built with a different compiler. As I updated gcc fairly recently I have a suspicion that the vcpkg dependencies have not been rebuilt since the compiler change. I wanted to quickly eliminate this possibility.

I'm using manifest mode with cmake (vcpkg CMAKE_TOOLCHAIN_FILE) on linux along with `vcpkg integrate install`. CMake is telling me it has detected the new compiler fine and everything seems to be properly configured on the cmake side.

**Reply by JackBoosY** [reply]
When gcc was updated, the port's abi info should be changed, then the related ports should be rebuild automaticly.

**Comment by romanholidaypancakes** [comment]
we need it(Automatically recompile all its child dependencies
```
.\vcpkg\vcpkg rebuild pkg 
```

**Reply by autoantwort** [reply]
Why do you need it? What is your goal?

**Reply by romanholidaypancakes** [reply]
Sometimes I change the compilation options or the source code, and then I have to recompile

**Reply by romanholidaypancakes** [reply]
@autoantwort I believe this type of operation is an everyday behavior in c++ projects

**Reply by dg0yt** [reply]
> @autoantwort I believe this type of operation is an everyday behavior in c++ projects

This may be true for the sources within projects.
But reinstalling the whole development environment is a much less frequent operation IMO.

**Comment by spinicist** [comment]
Hello,

I would very much like this functionality added.

My use case is that I installed the latest MacOS update last week and now every time I link my project I receive many, many, _many_ messages like this:

```
ld: warning: object file (/Users/tobias/Code/riesling/build/vcpkg_installed/arm64-osx/lib/libhdf5.a[195](H5Oattr.c.o)) was built for newer 'macOS' version (15.3) than being linked (15.0)
```

I'm talking many pages worth. Yes, I can change `CMAKE_OSX_DEPLOYMENT_TARGET` to 15.4, but I'd prefer to rebuild everything with the latest compiler.
